src/main/java/software/amazon/awssdk/crt/http/NonBlockingHttpResponsekHandler.java

```diff
@@ -0,0 +1,130 @@
+import software.amazon.awssdk.crt.utils.AwsCrtInputStreamEvent;
+import software.amazon.awssdk.crt.utils.AwsCrtPipedInputStream;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PipedOutputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Receives calls from Native code and translates native data to/from Java Objects
+ */
+class NonBlockingHttpResponsekHandler implements JniHttpCallbackHandler {
+    private final HttpRequest request;
+    private final CompletableFuture<HttpResponse> responseHandler;
+    private final HttpResponse.Builder responseBuilder = new HttpResponse.Builder();
+    private final PipedOutputStream out;
+    private final AwsCrtPipedInputStream in;
+
+    public NonBlockingHttpResponsekHandler(HttpRequest request, CompletableFuture<HttpResponse> responseHandler) throws IOException {
+        this.request = request;
+        this.responseHandler = responseHandler;
+        out = new PipedOutputStream();
+        AwsCrtInputStreamEvent callback = null; // TODO: wire up callback to update Window sizes on InputStream Read Events
+        in = new AwsCrtPipedInputStream(out, callback);
+    }
+
+    /**
+     * Called from Native when new Http Headers have been received.
+     * Note that this function may be called multiple times as HTTP headers are received.
+     * @param responseStatusCode The HTTP Response Status Code
+     * @param nextHeaders The headers received in the latest IO event.
+     */
+    public void onHeaders(HttpStream stream, int responseStatusCode, HttpHeader[] nextHeaders) {
+        responseBuilder.withStatusCode(responseStatusCode);
+        responseBuilder.withHeaders(nextHeaders);
+    }
+
+    /**
+     * Called from Native once all HTTP Headers are processed.
+     * @param hasBody True if the HTTP Response had a Body, false otherwise.
+     */
+    public void onHeadersDone(HttpStream stream, boolean hasBody) {
+        if (hasBody) {
+            responseBuilder.withBodyStream(in);
+        }
+        responseHandler.complete(responseBuilder.build());
+    }
+
+    /**
+     * Called when new Body bytes have been received.
+     * Note that this function may be called multiple times as headers are received.
+     * Do not keep a reference of the ByteBuffer after this method returns.
+     * @param bodyBytesIn The HTTP Body Bytes received in the last IO Event.
+     * @return The new IO Window Size
+     */
+    public void onResponseBody(HttpStream stream, ByteBuffer bodyBytesIn) {
+        if (!responseHandler.isDone()) {
+            /* May happen if there were no Headers in Response */
+            responseBuilder.withBodyStream(in);
+            responseHandler.complete(responseBuilder.build());
+        }
+
+        try {
+            /* Write the bytes, then update the Native Window Size*/
+            out.write(bodyBytesIn.array());
+            stream.updateWindowSize(in.writeSpaceAvailable());
```

Comment:
The "update" in update_window_size is a slightly misleading, but it's the terminology that's used in the world of read-windows...

This doesn't "set" the size of the window, it "re-opens" the window by that much.
So, by the fact that you've receive `bodyBytesIn` of data, the window has shrunk by `bodyBytesIn`.
You want to re-open the window by `bodyBytesIn`, if possible.

Long story short, this math should look something like: `stream.updateWindowSize(min(bodyBytesIn.length, in.writeSpaceAvailable());`

And if you didn't re-eopen the window all the way, you'll want to open it once space becomes available again